importlib_metadata/__init__.py

```diff
@@ -486,10 +495,35 @@ def _read_files_distinfo(self):
         text = self.read_text('RECORD')
         return text and text.splitlines()
 
-    def _read_files_egginfo(self):
+    def _read_files_egginfo_installed(self):
+        """
+        installed-files.txt might contain literal commas, so wrap
+        each line in quotes. Also, the entries in installed-files.txt
+        are relative to the .egg-info/ subdir (not relative to the
+        parent site-packages directory that make_file() expects).
+
+        This file is written when the package is installed by pip,
+        but it might not be written for other installation methods.
+        Hence, even if we can assume that this file is accurate
+        when it exists, we cannot assume that it always exists.
+        """
+        text = self.read_text('installed-files.txt')
+        # We need to prepend the .egg-info/ subdir to the lines in this file.
+        # But this subdir is only available in the PathDistribution's self._path
+        # which is not easily accessible from this base class...
+        subdir = getattr(self, '_path', None)
```

Comment:
This approach feels a little hacky. Perhaps it would make sense instead to prescribe an interface for `Distribution` objects to resolve paths.

Reply:
Yes, that is one solution. There is already `.locate_file()` which is not far off, except that its implementation in `PathDistribution` precisely _removes_ the metadata directory that we need in this case.

More generally, I wonder if there is just a tad too much coupling between the `Distribution` and `PathDistribution` classes?

There is a mix of abstract and concrete methods in the `Distribution` base class, and several of the concrete methods seem fairly tightly bound to the `PathDistribution` subclass. For example, the `at()` static method creates and returns a `PathDistribution` instance outright, and the `metadata()`, `entry_points()`, `files()`, and `requires()` methods are all implemented with assumptions as to which files should be available and that the organization should resemble a dist-info or egg-info distribution.

Now, there is (AFAICS) only the one (`PathDistribution`) subclass of `Distribution` available in this project, and I don't really have enough experience with this project or its users to really imagine what _other_ `Distribution` subclasses would or could exist in the wild.

Still, I would suspect that a subclass that was sufficiently different from `PathDistribution` to rather prefer subclassing `Distribution` directly, would then find itself not merely implementing the `read_text()` an `locate_file()` abstract methods, but would probably also need to _re_implement several of `metadata()`, `entry_points()`, `files()`, or `requires()` as well.

Hence, I would raise the question whether some of these methods would be better off with their concrete implementations moved into the `PathDistribution` subclass? (Of course leaving abstract methods behind in `Distribution` where that makes sense.) This would grant these concrete methods direct access to the stuff they need inside `PathDistribution`, instead of having to add more interfaces to `Distribution` for stuff that really only makes sense for `PathDistribution`.

Still, this is a much bigger refactoring than I set out to do, and I would not feel comfortable starting this without consulting you.

Reply:
> Now, there is (AFAICS) only the one (`PathDistribution`) subclass of `Distribution` available in this project, and I don't really have enough experience with this project or its users to really imagine what _other_ `Distribution` subclasses would or could exist in the wild.

This project does aim to support an abstract concept of a `Distribution`, which could be a package served from a database or binary bundle or dynamically from memory. One downstream derived type I've [encountered](https://github.com/brechtm/rinohtype/pull/270) is [DynamicRinohDistribution](https://github.com/jaraco/rinohtype/blob/e7f1c2c6066303d86cff4105be0829f512059be2/src/rinoh/resource.py#L132).

I think you're right that a review/refactor may be in order, but is probably out of scope for this effort. I'll re-use your analysis in a new issue to track follow-up work.